### PR TITLE
Add guide on integration test.

### PR DESCRIPTION
**Pull request by bianpengyuan**

### CODE REVIEW

**Comment by bianpengyuan** [comment]
@kyessenov ptal, thanks!

File: doc/write-integration-test.md [hunk 1]
```diff
@@ -1,3 +1,132 @@
...
 
-TODO
+Before deploying your Wasm extensions in production, it is **highly** recommended to write integration tests which load and run the extension with Envoy binary. To make writing such integration test easier, the integration test framework from istio/proxy repo will be used, which is the same framework used to test Istio telemetry extensions.
+
+## Overview
+---
+
+The reference document about the test framework could be found [here](https://godoc.org/github.com/istio/proxy/test/envoye2e/driver). At the high level, the test framework:
+* Downloads Envoy binary, which is built by Istio proxy postsubmit and used in istio proxy docker container.
+* Spawns up a xDS server locally, which serves customizable xDS resources. In your test, the XDS resource will point to your own Wasm extension.
```
Comment (kyessenov) [inline doc/write-integration-test.md:11]: _an xDS server_

File: doc/write-integration-test.md [hunk 2]
```diff
@@ -1,3 +1,132 @@
...
+* Spawns up Envoy processes locally with customizable bootstrap template.
```
Comment (kyessenov) [inline doc/write-integration-test.md:12]: _templates_

File: doc/write-integration-test.md [hunk 3]
```diff
@@ -1,3 +1,132 @@
...
+* Executes test logic which sends request through Envoy, then examines extension logic accordingly.
```
Comment (kyessenov) [inline doc/write-integration-test.md:13]: `requests`

File: doc/write-integration-test.md [hunk 4]
```diff
@@ -1,3 +1,132 @@
...
+
+To take a more detailed look, the test framework models every operation in the test as a [`Step`](https://godoc.org/github.com/istio/proxy/test/envoye2e/driver#Step). For example, one of the most important steps is [`Envoy`](https://godoc.org/github.com/istio/proxy/test/envoye2e/driver#Envoy), which downloads and spawns an Envoy process in `Run` method, and stops the process in `Clean up` method.
```
Comment (kyessenov) [inline doc/write-integration-test.md:15]: `CleanUp`

File: doc/write-integration-test.md [hunk 5]
```diff
@@ -1,3 +1,132 @@
...
+
+Every `Run` method implementation takes a [`Params`](https://godoc.org/github.com/istio/proxy/test/envoye2e/driver#Params) struct as input, which should be created at the beginning of the test and carries test context that is shared by all `Steps`. Information provided by `Params` including:
```
Comment (kyessenov) [inline doc/write-integration-test.md:17]: _includes_

File: doc/write-integration-test.md [hunk 6]
```diff
@@ -1,3 +1,132 @@
...
+* Port assignment, such as port assigned to client Envoy, server Envoy, and XDS server.
```
Comment (kyessenov) [inline doc/write-integration-test.md:18]: _port numbers assigned to the client Envoy, the server Envoy, and the xDS server_

File: doc/write-integration-test.md [hunk 7]
```diff
@@ -1,3 +1,132 @@
...
+* Variable map, which used to fill in templates for various Envoy configuration.
```
Comment (kyessenov) [inline doc/write-integration-test.md:19]: _is used_

File: doc/write-integration-test.md [hunk 8]
```diff
@@ -1,3 +1,132 @@
...
+
+## Example Extension Integration Test Walkthrough
+---
+
+As the first step of the test, a `Params` struct is created. Besides the Golang testing object, the other two variables are passed in:
+* a string map, which will be copied to `Vars` string map in `Params`. `Vars` are used to fill in all kinds of template that are used in the test, such as Envoy bootstrap configuration, listener configuration, filter configuration, etc. We will look at how `Vars` are used later.
+* a test inventory, which lists all tests that could be ran by go test. Inside the function, each test is going to have a unique set of ports assigned, which will be used later when spawn up Envoy and XDS server. Test inventory is to make sure that there is no port collision. For example, [here](https://github.com/istio/proxy/blob/32a5195862266bc49faa94bfb88d1719420abb3b/test/envoye2e/inventory.go#L21) is the test inventory of `istio/proxy`.
```
Comment (kyessenov) [inline doc/write-integration-test.md:29]: _prevents port collision_

**Review by kyessenov — APPROVED** [review]

**Pull request merged**